### PR TITLE
Add Weather Data File

### PR DESCRIPTION
**Pull request by NematMimi-22**

### CODE REVIEW

File: DataFormat/JSONWeatherData.cs [hunk 1]
```diff
@@ -0,0 +1,13 @@
+﻿using Newtonsoft.Json;
+namespace Weather_Monitoring.DataFormat
+{
+    public class JSONWeatherData : WeatherData
+    {
+        public static JSONWeatherData FromJson(string json)
+        {
+            var weatherData = JsonConvert.DeserializeObject<JSONWeatherData>(json);
+
+            return weatherData;
```
Comment (AmalDarKhalil) [inline DataFormat/JSONWeatherData.cs]: Please apply the Inline Temp that is taken in the refactoring course

File: DataFormat/WeatherDataFactory.cs [hunk 2]
```diff
@@ -0,0 +1,23 @@
...
+            else if (Regex.IsMatch(inputFormat.Trim(), @"^<.*>$"))
+            {
+                return XmlWeatherData.FromXml(inputFormat);
+            }
+            else
+            {
+                throw new ArgumentException("Unsupported input format");
+            }
+        }
+        }
```
Comment (AmalDarKhalil) [inline DataFormat/WeatherDataFactory.cs]: please fix the indentation

File: DataFormat/XmlWeatherData.cs [hunk 3]
```diff
@@ -0,0 +1,18 @@
...
+
+    public class XmlWeatherData : WeatherData
+    {
+        public static XmlWeatherData FromXml(string xml)
+        {
+            var serializer = new XmlSerializer(typeof(XmlWeatherData));
+            using (TextReader reader = new StringReader(xml))
+            {
+                var weatherData = (XmlWeatherData)serializer.Deserialize(reader);
+                return weatherData;
```
Comment (AmalDarKhalil) [inline DataFormat/XmlWeatherData.cs]: Please apply the Inline Temp that is taken in the refactoring course

File: Program.cs [hunk 4]
```diff
@@ -0,0 +1,35 @@
...
+        public static void Main()
+        {
+            var jsonFilePath = @"C:\Users\Nemat\source\repos\Weather-Monitoring\config.json";
+            var WeatherBots = ConfigReader.ReadConfig(jsonFilePath);
+
+            while (true)
+            {
+                Console.WriteLine("Enter the weather data:");
+                var input = Console.ReadLine();
+                var weatherData = WeatherDataFactory.CreateReader(input);
```
Comment (AmalDarKhalil) [inline Program.cs]: NIT: add a new line after this line

File: Program.cs [hunk 5]
```diff
@@ -0,0 +1,35 @@
...
+                if (WeatherBots.RainBot.enabled=true && WeatherBots.RainBot.IsActivated(weatherData))
+                {
+                    WeatherBots.RainBot.PerformAction();
+                    Console.WriteLine($"RainBot:{WeatherBots.RainBot.message}");
+                }
```
Comment (AmalDarKhalil) [inline Program.cs]: NIT: add a new line after this line

File: Program.cs [hunk 6]
```diff
@@ -0,0 +1,35 @@
...
+                if (WeatherBots.SnowBot.enabled = true && WeatherBots.SnowBot.IsActivated(weatherData))
+                {
+                    WeatherBots.SnowBot.PerformAction();
+                    Console.WriteLine($"SnowBot:{WeatherBots.SnowBot.message}");
+                }
```
Comment (AmalDarKhalil) [inline Program.cs:50]: NIT: add a new line after this line

File: Bots/SnowBot.cs [hunk 7]
```diff
@@ -0,0 +1,17 @@
+﻿using Weather_Monitoring.DataFormat;
+namespace Weather_Monitoring.Bots
+{
+    public class SnowBot : BaseBot
+    {
+        public double TemperatureThreshold { get; set; } 
+        public override bool IsActivated(WeatherData weatherData)
+        {
+            return weatherData.Temperature < TemperatureThreshold;
+        }
```
Comment (AmalDarKhalil) [inline Bots/SnowBot.cs]: this should be an extension method, no?

File: Bots/ConfigReader.cs [hunk 8]
```diff
@@ -0,0 +1,14 @@
+﻿using System.Text.Json;
+namespace Weather_Monitoring.Bots
+{
+    public class ConfigReader
+    {
+        public static WeatherBots ReadConfig(string configFile)
+        {
+            string jsonContent = File.ReadAllText(configFile);
```
Comment (AmalDarKhalil) [inline Bots/ConfigReader.cs]:
```suggestion
            var jsonContent = File.ReadAllText(configFile);
```

File: Bots/RainBot.cs [hunk 9]
```diff
@@ -0,0 +1,16 @@
+﻿using Weather_Monitoring.DataFormat;
+namespace Weather_Monitoring.Bots
+{
+    public class RainBot : BaseBot
+    {
+        public int HumidityThreshold { get; set; }
```
Comment (AmalDarKhalil) [inline Bots/RainBot.cs:6]: NIT
```suggestion
        public int HumidityThreshold { get; set; }
        
```

File: Bots/RainBot.cs [hunk 10]
```diff
@@ -0,0 +1,16 @@
...
+        public bool IsActivated(WeatherData weatherData)
+        {
+            return weatherData.Humidity > HumidityThreshold;
+        }
```
Comment (AmalDarKhalil) [inline Bots/RainBot.cs:17]: NIT
```suggestion
        }
        
```

Comment (AmalDarKhalil) [reply Bots/RainBot.cs:17]: same for SnowBot  and SunBot

File: Bots/RainBot.cs (hunk 9 above)
Comment (AmalDarKhalil) [reply Bots/RainBot.cs:6]: same for SnowBot  and SunBot

File: DataFormat/XmlWeatherData.cs [hunk 11]
```diff
@@ -0,0 +1,17 @@
+﻿using System.Xml.Serialization;
+namespace Weather_Monitoring.DataFormat
+{
+    [XmlRoot("WeatherData", Namespace = "")]
+
+    public class XmlWeatherData : WeatherData
+    {
+        public static XmlWeatherData FromXml(string xml)
+        {
+            var serializer = new XmlSerializer(typeof(XmlWeatherData));
```
Comment (AmalDarKhalil) [inline DataFormat/XmlWeatherData.cs]: NIT
```suggestion
            var serializer = new XmlSerializer(typeof(XmlWeatherData));
            
```

File: Bots/ConfigReader.cs [hunk 12]
```diff
@@ -0,0 +1,14 @@
...
+            WeatherBots WeatherBots = JsonSerializer.Deserialize<WeatherBots>(jsonContent);
```
Comment (AmalDarKhalil) [inline Bots/ConfigReader.cs]:
```suggestion
            var WeatherBots = JsonSerializer.Deserialize<WeatherBots>(jsonContent);
```

File: Program.cs [hunk 13]
```diff
@@ -0,0 +1,42 @@
+﻿using Weather_Monitoring.Bots;
+using Weather_Monitoring.DataFormat;
+using Weather_Monitoring.Enums;
+namespace Weather_Monitoring
+{
+    public class Program
+    {
+        public static void Main()
+        {
+            var jsonFilePath = @"C:\Users\Nemat\source\repos\Weather-Monitoring\config.json";
```
Comment (AmalDarKhalil) [inline Program.cs]: please make it dynamic path, input from the user

File: Program.cs [hunk 14]
```diff
@@ -0,0 +1,42 @@
...
+            while (true)
+            {
+                Console.WriteLine("Enter the weather data:");
+                var input = Console.ReadLine();
+                var weatherData = WeatherDataFactory.CreateReader(input);
+                if (WeatherBots.RainBot.enabled && WeatherBots.RainBot.IsActivated(weatherData))
+                {
+                    var Rainbot = basebotFactory.GetBot(BotType.RainBot, WeatherBots);
+                    Rainbot.PerformAction();
+                    Console.WriteLine($"{Rainbot.Name}:{Rainbot.message}");
```
Comment (AmalDarKhalil) [inline Program.cs]: create a printer service

File: Bots/BaseBotFactory.cs [hunk 15]
```diff
@@ -0,0 +1,21 @@
+﻿using Weather_Monitoring.Enums;
+namespace Weather_Monitoring.Bots
+{
+    public class BaseBotFactory
+    {
+        public IBaseBot GetBot(BotType botType, WeatherBots weatherBots)
```
Comment (AmalDarKhalil) [inline Bots/BaseBotFactory.cs]: this should be abstract, no?

Comment (AmalDarKhalil) [reply Bots/BaseBotFactory.cs]: please check this example
https://refactoring.guru/design-patterns/abstract-factory/csharp/example#example-0

File: Program.cs [hunk 16]
```diff
@@ -0,0 +1,42 @@
...
+                    Snowbot.PerformAction();
+                    Console.WriteLine($"{Snowbot.Name}:{Snowbot.message}");
+                }
+
+                if (WeatherBots.SunBot.enabled && WeatherBots.SunBot.IsActivated(weatherData))
+                {
+                    var Sunbot = basebotFactory.GetBot(BotType.SunBot, WeatherBots);
+                    Sunbot.PerformAction();
+                    Console.WriteLine($"{Sunbot.Name}:{Sunbot.message}");
+                }
```
Comment (AmalDarKhalil) [inline Program.cs]: every bot is IBaseBot, right? why not use the base bot to execute PerformAction()?

**Review by AmalDarKhalil — APPROVED** [review]

File: Program.cs [hunk 17]
```diff
@@ -0,0 +1,45 @@
...
+using Weather_Monitoring.PrinterService;
+using Weather_Monitoring.Factory;
+namespace Weather_Monitoring
+{
+    public class Program
+    {
+        public static void Main()
+        {
+            Console.Write("Enter the path to the JSON config file: ");
+            var jsonFilePath = Console.ReadLine();
```
Comment (kshawish) [inline Program.cs]: there's a better way to handle configuration in C#/.Net  do some research about that.

File: Program.cs [hunk 18]
```diff
@@ -0,0 +1,45 @@
...
+            var WeatherBots = ConfigReader.ReadConfig(jsonFilePath);
```
Comment (kshawish) [inline Program.cs]: it's strange that ConfigReader.ReadConfig is creating the real bots, I'd expect it to only read the configuration that would be used to create and configure the bots. that way that you did it couples the bot Configuration to the bots themselves.

Comment (kshawish) [inline Program.cs]: camelCase for local variables, read more:
https://learn.microsoft.com/en-us/dotnet/csharp/fundamentals/coding-style/coding-conventions

File: Program.cs [hunk 19]
```diff
@@ -0,0 +1,45 @@
...
+            var basebotFactory = new BaseBotFactory();
+
+            while (true)
+            {
+                Console.WriteLine("Enter the weather data:");
```
Comment (kshawish) [inline Program.cs:27]: you should've used the pub/sub pattern here, that would make it a lot easier to extend the app and add more bots and subscriber for  weather data

File: Program.cs [hunk 20]
```diff
@@ -0,0 +1,45 @@
...
+                var input = Console.ReadLine();
+                var weatherData = WeatherDataFactory.CreateReader(input);
+                if (WeatherBots.RainBot.enabled && WeatherBots.RainBot.IsActivated(weatherData))
+                {
+                    var Rainbot = basebotFactory.GetBot(BotType.RainBot, WeatherBots);
```
Comment (kshawish) [inline Program.cs]: that's creating a bot each time we read data !! what if the bot object was a heavy one?!

File: Program.cs [hunk 21]
```diff
@@ -0,0 +1,45 @@
...
+                }
+
+                if (WeatherBots.SnowBot.enabled && WeatherBots.SnowBot.IsActivated(weatherData))
+                {
+                    var Snowbot = basebotFactory.GetBot(BotType.SnowBot, WeatherBots);
+                    Snowbot.PerformAction();
+                    BotMessagePrinter.PrintBotMessage(Snowbot);
+                }
+
+                if (WeatherBots.SunBot.enabled && WeatherBots.SunBot.IsActivated(weatherData))
```
Comment (kshawish) [inline Program.cs]: The bot shouldn't expose a method called IsActivated, will explain why.

File: Program.cs [hunk 22]
```diff
@@ -0,0 +1,45 @@
...
+                {
+                    var Sunbot = basebotFactory.GetBot(BotType.SunBot, WeatherBots);
+                    Sunbot.PerformAction();
+                    BotMessagePrinter.PrintBotMessage(Sunbot);
```
Comment (kshawish) [inline Program.cs]: isn't PerformAction responsible to print the message ?!

File: PrinterService/BotMessagePrinter.cs [hunk 23]
```diff
@@ -0,0 +1,11 @@
+﻿using Weather_Monitoring.Bots;
+namespace Weather_Monitoring.PrinterService
+{
+    public class BotMessagePrinter
```
Comment (kshawish) [inline PrinterService/BotMessagePrinter.cs]: not needed.

File: Factory/SunBotFactory.cs [hunk 24]
```diff
@@ -0,0 +1,11 @@
+﻿using Weather_Monitoring.Bots;
+namespace Weather_Monitoring.Factory
+{
+    public class SunBotFactory : IWeatherBotFactory
```
Comment (kshawish) [inline Factory/SunBotFactory.cs]: I don't this is a correct use for the factory pattern. would explain

File: Factory/RainBotFactory.cs [hunk 25]
```diff
@@ -0,0 +1,11 @@
+﻿using Weather_Monitoring.Bots;
+namespace Weather_Monitoring.Factory
+{
+    public class RainBotFactory : IWeatherBotFactory
+    {
+        public IBaseBot CreateBot(WeatherBots weatherBots)
+        {
+            return new RainBot(weatherBots.RainBot.HumidityThreshold, weatherBots.RainBot.enabled, weatherBots.RainBot.message);
```
Comment (kshawish) [inline Factory/RainBotFactory.cs]: it's strange to depend  on a RainBot ( weatherBots.RainBot) in a factory of RainBots to create a RainBot.

File: DataFormat/XmlWeatherData.cs [hunk 26]
```diff
@@ -0,0 +1,18 @@
+﻿using System.Xml.Serialization;
+namespace Weather_Monitoring.DataFormat
+{
+    [XmlRoot("WeatherData", Namespace = "")]
+
+    public class XmlWeatherData : WeatherData
```
Comment (kshawish) [inline DataFormat/XmlWeatherData.cs]: you should've separated the data and the behavior. will explain.

File: DataFormat/XmlWeatherData.cs [hunk 27]
```diff
@@ -0,0 +1,18 @@
...
+    {
+        public static XmlWeatherData FromXml(string xml)
+        {
+            var serializer = new XmlSerializer(typeof(XmlWeatherData));
+
+            using (TextReader reader = new StringReader(xml))
+            {
+                return (XmlWeatherData)serializer.Deserialize(reader);
```
Comment (kshawish) [inline DataFormat/XmlWeatherData.cs]: this should retrun just a weather data

File: DataFormat/WeatherDataFactory.cs [hunk 28]
```diff
@@ -0,0 +1,23 @@
+﻿using System.Text.RegularExpressions;
+
+namespace Weather_Monitoring.DataFormat
+{
+    public class WeatherDataFactory
+    {
+        public static WeatherData CreateReader(string inputFormat)
+        {
+            if (Regex.IsMatch(inputFormat.Trim(), @"^{.*}$"))
```
Comment (kshawish) [inline DataFormat/WeatherDataFactory.cs]: encapsulate this logic 'Regex.IsMatch(inputFormat.Trim(), @"^{.*}$")' inside  IsJson function.

File: DataFormat/JSONWeatherData.cs [hunk 29]
```diff
@@ -0,0 +1,11 @@
+﻿using Newtonsoft.Json;
+namespace Weather_Monitoring.DataFormat
+{
+    public class JSONWeatherData : WeatherData
```
Comment (kshawish) [inline DataFormat/JSONWeatherData.cs]: nop.  same comment for xmlWetherData

File: Bots/IBaseBot.cs [hunk 30]
```diff
@@ -0,0 +1,14 @@
+﻿using Weather_Monitoring.DataFormat;
+namespace Weather_Monitoring.Bots
+{
+    public interface IBaseBot
```
Comment (kshawish) [inline Bots/IBaseBot.cs]: because it's just an internface call it just IBot.

File: ReadConfig/ConfigReader.cs [hunk 31]
```diff
@@ -0,0 +1,31 @@
+﻿using Microsoft.Extensions.Configuration;
+namespace Weather_Monitoring.ReadConfig
+{
+    public class ConfigReader
+    {
+        public static BotConfig ReadConfig(string botName)
+        {
+            IConfiguration configuration = new ConfigurationBuilder()
+           .SetBasePath(Directory.GetCurrentDirectory())
+           .AddJsonFile("C:\\Users\\Nemat\\source\\repos\\Weather-Monitoring\\ReadConfig\\config.json")
```
Comment (AmalDarKhalil) [inline ReadConfig/ConfigReader.cs]: can't we depend on the relative path instead of the absolute one?

File: ReadConfig/ConfigReader.cs [hunk 32]
```diff
@@ -0,0 +1,31 @@
...
+           .Build();
+            var Threshold = GetThreshold(botName, configuration);
+            var Message= configuration[$"{botName}:Message"];
```
Comment (AmalDarKhalil) [inline ReadConfig/ConfigReader.cs]:
```suggestion
            var Message = configuration[$"{botName}:Message"];
```

File: ReadConfig/ConfigReader.cs [hunk 33]
```diff
@@ -0,0 +1,31 @@
...
+        private static int GetThreshold(string botName, IConfiguration configuration)
+        {
+            if (botName == "RainBot")
+            {
+                return int.Parse(configuration["RainBot:HumidityThreshold"]);
+            }
+            else
+            {
+                return int.Parse(configuration["SunBot:TemperatureThreshold"]);
+            }
```
Comment (AmalDarKhalil) [inline ReadConfig/ConfigReader.cs:32]: snow bot?

File: PublishSubscribeDP/WeatherEventSubscriber.cs [hunk 34]
```diff
@@ -0,0 +1,18 @@
...
+{
+    public class WeatherEventSubscriber
+    {
+        public static Dictionary<string, List<IBot>> subscribers = new Dictionary<string, List<IBot>>();
+
+        public void Subscribe(string eventName, IBot bot)
+        {
+            if (!subscribers.ContainsKey(eventName))
+            {
+
```
Comment (AmalDarKhalil) [inline PublishSubscribeDP/WeatherEventSubscriber.cs:12]: extra line

File: PublishSubscribeDP/WeatherEventPublisher.cs [hunk 35]
```diff
@@ -0,0 +1,17 @@
+﻿using Weather_Monitoring.DataFormat;
+namespace Weather_Monitoring.PublishSubscribeDP
+{
+    public class WeatherEventPublisher
+    {
+        public void PublishEvent(string eventName, WeatherData weatherData)
+        {
+            if (WeatherEventSubscriber.subscribers.ContainsKey(eventName))
```
Comment (AmalDarKhalil) [inline PublishSubscribeDP/WeatherEventPublisher.cs]: what about using TryGetValue?

File: ReadConfig/ConfigReader.cs [hunk 36]
```diff
@@ -0,0 +1,35 @@
...
+        {
+            if (botName == "RainBot")
+            {
+                return int.Parse(configuration["RainBot:HumidityThreshold"]);
+            }
+            else if (botName == "SnowBot")
+            {
+                return int.Parse(configuration["SnowBot:TemperatureThreshold"]);
+            }
+            else
```
Comment (AmalDarKhalil) [inline ReadConfig/ConfigReader.cs]: do we need to type else in this case?

File: Program.cs [hunk 37]
```diff
@@ -0,0 +1,33 @@
...
+            while (true)
+            {
+                Console.WriteLine("Enter the weather data:");
+                var input = Console.ReadLine();
+                var weatherData = WeatherDataFactory.CreateReader(input);
+                var eventPublisher = new WeatherEventPublisher();
+                eventPublisher.PublishEvent("HumidityExceeded", weatherData);
+                eventPublisher.PublishEvent("TemperatureExceeded", weatherData);
+                eventPublisher.PublishEvent("TemperatureDropped", weatherData);
+            }
```
Comment (AmalDarKhalil) [inline Program.cs:55]: Shouldn't we unsubscribe as well?

File: PublishSubscribeDP/WeatherEventPublisher.cs [hunk 38]
```diff
@@ -0,0 +1,17 @@
+﻿using Weather_Monitoring.DataFormat;
+namespace Weather_Monitoring.PublishSubscribeDP
+{
+    public class WeatherEventPublisher
+    {
+        public void PublishEvent(string eventName, WeatherData weatherData)
+        {
+            if (WeatherEventSubscriber.subscribers.TryGetValue(eventName, out var subscribers))
+            {
+                foreach (var bot in WeatherEventSubscriber.subscribers[eventName])
```
Comment (AmalDarKhalil) [inline PublishSubscribeDP/WeatherEventPublisher.cs]:
```suggestion
                foreach (var bot in subscribers)
```
that's why I suggest TryGetValue

File: DataFormat/WeatherDataFactory.cs [hunk 39]
```diff
@@ -0,0 +1,33 @@
...
+        {
+            if (IsJSONFormat(inputFormat))
+            {
+                return JSONParser.FromJson(inputFormat);
+            }
+            else if (IsXMLFormat(inputFormat))
+            {
+                return XmlParser.FromXml(inputFormat);
+            }
+            else
```
Comment (AmalDarKhalil) [inline DataFormat/WeatherDataFactory.cs]: do we need else?

File: DataFormat/WeatherDataFactory.cs [hunk 40]
```diff
@@ -0,0 +1,33 @@
...
+{
+    public class WeatherDataFactory
+    {
+        public static WeatherData CreateReader(string inputFormat)
+        {
+            if (IsJSONFormat(inputFormat))
+            {
+                return JSONParser.FromJson(inputFormat);
+            }
+            else if (IsXMLFormat(inputFormat))
```
Comment (AmalDarKhalil) [inline DataFormat/WeatherDataFactory.cs]: do we need else?

File: ReadConfig/ConfigReader.cs [hunk 41]
```diff
@@ -0,0 +1,36 @@
...
+            return new BotConfig(Enabled, Threshold, Message); 
+        }
+
+        private static int GetThreshold(string botName, IConfiguration configuration)
+        {
+            if (botName == "RainBot")
+            {
+                return int.Parse(configuration["RainBot:HumidityThreshold"]);
+            }
+            else if (botName == "SnowBot")
```
Comment (AmalDarKhalil) [inline ReadConfig/ConfigReader.cs]: do we need else?

**Pull request merged**